File: futpools_web/src/lib/prizeLadder.js
// Prize-ladder helpers for the web client — mirrors
// futpools_backend/src/lib/prizeLadder.js. Keep the two in sync.
//
// A ladder is an array of range tiers { min, max, prizeMXN, label? }:
// "if your correct-pick count is in [min, max], you win prizeMXN pesos".
// Ranges are NOT assumed monotonic (0 aciertos → $100 consolation sits
// below the 1–7 dead zone), so lookups never rely on ordering.

export const DEFAULT_LADDER = [
  { min: 12, max: 12, prizeMXN: 3000 },
  { min: 11, max: 11, prizeMXN: 1500 },
  { min: 10, max: 10, prizeMXN: 700 },
  { min: 9, max: 9, prizeMXN: 300 },
  { min: 8, max: 8, prizeMXN: 100 },
  { min: 1, max: 7, prizeMXN: 0 },
  { min: 0, max: 0, prizeMXN: 100 },
];

function toInt(n) {
  const v = Number(n);
  return Number.isFinite(v) ? Math.trunc(v) : NaN;
}

/** Prize (MXN) for a given number of correct picks; 0 if no tier matches. */
export function prizeForCorrect(ladder, correctCount) {
  if (!Array.isArray(ladder)) return 0;
  const c = toInt(correctCount);
  if (Number.isNaN(c)) return 0;
  for (const tier of ladder) {
    const min = toInt(tier?.min);
    const max = toInt(tier?.max);
    if (Number.isNaN(min) || Number.isNaN(max)) continue;
    if (c >= min && c <= max) {
      const prize = Number(tier.prizeMXN);
      return Number.isFinite(prize) ? prize : 0;
    }
  }
  return 0;
}

/**
 * Closest paying tier above the current correct count — what the player is
 * chasing. Returns { tier, needed } or null when there's nothing higher.
 */
export function nextPrizeTier(ladder, currentCorrect) {
  if (!Array.isArray(ladder)) return null;
  const c = toInt(currentCorrect);
  if (Number.isNaN(c)) return null;
  let best = null;
  for (const tier of ladder) {
    const min = toInt(tier?.min);
    const prize = Number(tier?.prizeMXN);
    if (Number.isNaN(min) || !Number.isFinite(prize) || prize <= 0) continue;
    if (min > c && (best === null || min < toInt(best.min))) best = tier;
  }
  if (!best) return null;
  return { tier: best, needed: toInt(best.min) - c };
}

/** The biggest prize on the ladder (the headline "win up to" figure). */
export function maxPrize(ladder) {
  if (!Array.isArray(ladder)) return 0;
  return ladder.reduce((m, tier) => Math.max(m, Number(tier?.prizeMXN) || 0), 0);
}

/** Top paying tiers, highest prize first — for compact ladder previews. */
export function topTiers(ladder, n = 4) {
  if (!Array.isArray(ladder)) return [];
  return ladder
    .filter((tier) => (Number(tier?.prizeMXN) || 0) > 0)
    .sort((a, b) => (b.prizeMXN - a.prizeMXN) || (b.max - a.max))
    .slice(0, n);
}

/** Format MXN with thousands separators, e.g. 3000 → "$3,000". */
export function formatMXN(amount) {
  const n = Number(amount) || 0;
  return `$${n.toLocaleString('en-US')}`;
}

/**
 * Human label for a tier's acierto range, localized.
 * tierRangeLabel({min:1,max:7}, 'es') → "1–7 aciertos"
 * tierRangeLabel({min:10,max:10}, 'en') → "10 correct"
 */
export function tierRangeLabel(tier, locale = 'es') {
  const min = toInt(tier?.min);
  const max = toInt(tier?.max);
  const word = locale === 'es' ? 'aciertos' : 'correct';
  if (Number.isNaN(min) || Number.isNaN(max)) return '';
  return min === max ? `${min} ${word}` : `${min}–${max} ${word}`;
}

/** Full prize label used in notifications, e.g. "$700 (10 aciertos)". */
export function tierPrizeLabel(tier, locale = 'es') {
  return `${formatMXN(tier?.prizeMXN)} (${tierRangeLabel(tier, locale)})`;
}

File: futpoolsapp/futpoolsapp/Models/PrizeLadder.swift
//
//  PrizeLadder.swift
//  futpoolsapp
//
//  Prize-ladder model + helpers for the `prize_ladder` pool type, where
//  each participant wins a fixed prize based on their number of correct
//  picks ("aciertos"). Mirrors futpools_backend/src/lib/prizeLadder.js
//  and futpools_web/src/lib/prizeLadder.js — keep the three in sync.
//

import Foundation

/// One rung of a prize ladder: "if your correct-pick count is in
/// [min, max], you win prizeMXN pesos". Ranges are NOT assumed monotonic
/// (the 0-aciertos consolation sits below the 1–7 dead zone).
struct PrizeTier: Decodable, Identifiable, Equatable {
    let min: Int
    let max: Int
    let prizeMXN: Int
    let label: String?

    var id: String { "\(min)-\(max)" }

    enum CodingKeys: String, CodingKey { case min, max, prizeMXN, label }

    init(min: Int, max: Int, prizeMXN: Int, label: String? = nil) {
        self.min = min
        self.max = max
        self.prizeMXN = prizeMXN
        self.label = label
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        min = try c.decodeIfPresent(Int.self, forKey: .min) ?? 0
        max = try c.decodeIfPresent(Int.self, forKey: .max) ?? 0
        prizeMXN = try c.decodeIfPresent(Int.self, forKey: .prizeMXN) ?? 0
        label = try c.decodeIfPresent(String.self, forKey: .label)
    }
}

enum PrizeLadder {
    /// Prize (MXN) for a given number of correct picks; 0 if no tier matches.
    static func prize(for correct: Int, in ladder: [PrizeTier]) -> Int {
        for tier in ladder where correct >= tier.min && correct <= tier.max {
            return tier.prizeMXN
        }
        return 0
    }

    /// The tier the player currently sits in (for highlighting).
    static func currentTier(for correct: Int, in ladder: [PrizeTier]) -> PrizeTier? {
        ladder.first { correct >= $0.min && correct <= $0.max }
    }

    /// Closest *paying* tier above the current correct count — what the
    /// player is chasing. Returns (tier, needed) or nil if nothing higher.
    static func nextPrizeTier(for correct: Int, in ladder: [PrizeTier]) -> (tier: PrizeTier, needed: Int)? {
        var best: PrizeTier?
        for tier in ladder where tier.prizeMXN > 0 && tier.min > correct {
            if best == nil || tier.min < best!.min { best = tier }
        }
        guard let best else { return nil }
        return (best, best.min - correct)
    }

    /// "$3,000" formatting with thousands separators.
    static func formatMXN(_ amount: Int) -> String {
        let fmt = NumberFormatter()
        fmt.numberStyle = .decimal
        fmt.groupingSeparator = ","
        let n = fmt.string(from: NSNumber(value: amount)) ?? "\(amount)"
        return "$\(n)"
    }

    /// Localized "10 aciertos" / "1–7 aciertos" range label.
    static func rangeLabel(_ tier: PrizeTier) -> String {
        let word = String(localized: "aciertos_word")
        return tier.min == tier.max
            ? "\(tier.min) \(word)"
            : "\(tier.min)–\(tier.max) \(word)"
    }

    /// Full prize label used in notifications, e.g. "$700 (10 aciertos)".
    static func prizeLabel(_ tier: PrizeTier) -> String {
        "\(formatMXN(tier.prizeMXN)) (\(rangeLabel(tier)))"
    }
}
